### ISSUE TITLE
Typescript does not detect interface attributes when returning union type

### ISSUE THREAD

**Issue by georg94** [Microsoft/TypeScript#24103]
**TypeScript Version 2.9.0-dev.20180512

**Search Terms:** 
* Interface union type
* return union type interface and generic

**Code**

```ts
interface AddressInfo {
	port: number;
	host: string;
}

function getAddress(): AddressInfo | string {
	return {
		port: 1234,
		host: 'localhost'
	};
}

console.log(getAddress().port);
```

**Expected behavior:**
No errors should occur.

**Actual behavior:**
The Typescript Transpiler says `Property 'port' does not exist on type 'string'.` although it is defined on the interface and set as union for the retrun type. Only using `AddressInfo` works.

**Playground Link:**  http://www.typescriptlang.org/play/#src=interface%20AddressInfo%20%7B%0D%0A%09port%3A%20number%3B%0D%0A%09host%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Afunction%20getAddress()%3A%20AddressInfo%20%7C%20string%20%7B%0D%0A%09return%20%7B%0D%0A%09%09port%3A%201234%2C%0D%0A%09%09host%3A%20'localhost'%0D%0A%09%7D%3B%0D%0A%7D%0D%0A%0D%0Aconsole.log(getAddress().port)%3B%0D%0A

**Related Issues:**
